docs/RFCS/20200810_leaseholder_cache_invalidation.md

```diff
@@ -0,0 +1,464 @@
+
+- Like above, there's a delay between the moment a node terminates
+  ([step 5 of drain](#draining)) and the moment other nodes learn of
+  the new lease records. During that time, requests can be mis-routed.
+
+  Like above, this last cons can be alleviated by extending step 5 of
+  the draining process by a "sleep period".
+
+### Multi-tenant variant
+
+- The KV servers receive the "lease records" over gossip from other KV servers.
+- The KV servers *forward* the lease records to SQL tenants,
+  where the range caches are stored.
+
+## 4. Liveness-based cache invalidation with poison
+
+Outline:
+
+- [liveness updates](#liveness-updates) are gossiped as usual.
+- upon receiving a liveness update that indicates that another node is
+  not live, each node updates its range cache as follows:
+
+  - invalidate all leases pointing to the now-unavailable node, ie.
+    remove the lease from the cache.
+  - mark that store/node ID as "not preferred" for the subsequent lease
+    discovery. This *“poisons”* that store/node ID.
+
+- upon a cache miss, the range cache will still do lease discovery,
+  but it will skip over "not preferred" nodes/stores in the first
+  round.
+
+  Only if a lease is not discovered during the first round, are
+  poisoned nodes added to the lookup in the second and next rounds.
+
+**Pros:**
+
+- Addresses both [scenarios A and B](#scenarios).
+- By using gossip, the data will eventually reach every other
+  node even those that were temporarily unavailable.
+  (i.e. eliminates one cons of solution 1 above)
+- Does not add new data in gossip. Does not violate the [design
+  principles](#general-cockroachdb-design-principles).
+
+**Cons:**
+
+- Added complexity in the range cache to mark node/store IDs as
+  poisoned.
+- Added complexity in the lease discovery process.
+- There is still a need for lease discovery; there is a chance
+  that the first next replica contacted during discovery
+  will take a while to respond.
+- Like above, there's a delay between the moment a node terminates
+  ([step 5 of drain](#draining)) and the moment other nodes learn of
+  the new lease records. During that time, requests can be mis-routed.
+
+  Like above, this last cons can be alleviated by extending step 5 of
+  the draining process by a "sleep period".
+
+### Multi-tenant variant
+
+- The KV servers receive the liveness updates over gossip from other
+  KV servers.
+- The KV servers *forward* the liveness updates to SQL tenants,
+  where the range caches are stored.
+
+## 5. Liveness-based cache invalidation with bypass
+
+Outline:
+
+- [liveness updates](#liveness-updates) are gossiped as usual.
+- nothing particular happens upon receiving a liveness update
+  (other than updating the liveness cache as usual).
+- upon a cache lookup, the range cache consults liveness.
+
+  If the entry in the cache point to a node for which liveness
+  currently says the node is not live, the cache entry is ignored (is
```

Comment:
I think we should do some lip service around here to the fact that we can determine separately whether a lease is stale (its epoch is lower than the corresponding liveness record's liveness) and whether a node's expiration record is expired. I'm don't think the first info is very useful - everything else being equal, the stale leaseholder node is still the most likely node to be the leaseholder I think (e.g. if it had a liveness blip but everybody behaves like today, it'll get all its leases back).
A case that I think deserve thought is a valid lease on a dead node (so, an expiration-based lease on a node with an expired liveness record). 

Btw there's another source of liveness info that we sometimes use - the connection health exposed by the `RPCContext`; I don't know if you want to bring it into this discussion though.